include/zephyr/location.h

```diff
@@ -0,0 +1,340 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Location data
+ *
+ * The latitude and longitude are in the unit micro degree. 1 micro
+ * degree corresponds to approx. 11cm along the earths surface.
+ *
+ * The accuracy describes the radius within which the real
+ * position is with a 68.2% certainty, in millimeters.
+ */
+struct location_position {
+	/** Latitudal position in microdegrees (0 to +-180000000) */
+	int32_t latitude;
+	/** Longitudal position in microdegrees (0 to +-180000000) */
+	int32_t longitude;
+	/** The uptime when the position was determined */
+	int64_t uptime;
+	/** Accuracy radius in millimeters */
+	uint32_t accuracy;
+};
+
+/**
+ * @brief Bearing data
+ *
+ * The accuracy describes the uncertainty of the bearing in +-
+ * millidegrees with a 68.2% certainty.
+ */
+struct location_bearing {
+	/** Bearing angle relative to True North in millidegrees (0 to 360000) */
+	uint32_t bearing;
+	/** The uptime when the bearing was determined */
+	int64_t uptime;
+	/** Accuracy in millidegrees */
+	uint32_t accuracy;
+};
+
+/**
+ * @brief Speed data
+ *
+ * The accuracy describes the uncertainty of the speed in +-
+ * millimeters pr second with a 68.2% certainty.
+ */
+struct location_speed {
+	/** Speed over ground in millimeters pr second */
+	uint32_t speed;
+	/** The uptime when the speed was determined */
+	int64_t uptime;
+	/** Accuracy in millimeters pr second */
+	uint32_t accuracy;
+};
+
+/**
+ * @brief Altitude data
+ *
+ * The accuracy describes the uncertainty of the altitude
+ * in +- millimeters with a 68.2% certainty.
+ */
+struct location_altitude {
+	/** Altitude above mean sea level in millimeters */
+	int32_t altitude;
+	/** The uptime when the altitude was determined */
+	int64_t uptime;
+	/** Accuracy in millimeters */
+	uint32_t accuracy;
+};
+
+/**
+ * @typedef location_position_get_t
+ * @brief API for getting newest position data
+ *
+ * See struct location_position for description of position data
+ */
+typedef int (*location_position_get_t)(const struct device *dev,
+				       struct location_position *position);
+
+/**
+ * @typedef location_bearing_get_t
+ * @brief API for getting newest bearing data
+ *
+ * See struct location_bearing for description of bearing data
+ */
+typedef int (*location_bearing_get_t)(const struct device *dev,
+				      struct location_bearing *bearing);
+
+/**
+ * @typedef location_speed_get_t
+ * @brief API for getting newest speed data
+ *
+ * See struct location_speed for description of speed data
+ */
+typedef int (*location_speed_get_t)(const struct device *dev,
+				    struct location_speed *speed);
+
+/**
+ * @typedef location_altitude_get_t
+ * @brief API for getting newest altitude data
+ *
+ * See struct location_altitude for description of altitude data
+ */
+typedef int (*location_altitude_get_t)(const struct device *dev,
+				       struct location_altitude *altitude);
+
+/**
+ * @brief Location API
+ */
+__subsystem struct location_provider_api {
```

Comment:
Is this sustainable when we implement more info of the NMEA sentence in the future? Each of the data seems to have its own `uptime` and `accuracy` field, I'm wondering if it would be easier if they are in a single struct instead?

Reply:
The location API is not tied directly with NMEA frames, the GNSS driver will handle these.

The sustainability of this API/subsystem is a result of the location data only being the essentials provided by any location provider. As long as the earth is spherical, and physics don't change, the API is essentially "future proof".

Technically we can combine the uptime and accuracy into a struct and name it something like metadata. I don't know if it is actually a useful idea though, I personally prefer the verbose nature of having the fields there.